File: scala/protein-translation/src/main/scala/ProteinTranslation.scala
object ProteinTranslation {
  def proteins(rna: String) =
    rna.grouped(3).map(codonToProtein).takeWhile(_ != "STOP").toList

  private def codonToProtein(codon: String) =
    codon match {
      case "AUG" => "Methionine"
      case "UUC" => "Phenylalanine"
      case "UUU" => "Phenylalanine"
      case "UUA" => "Leucine"
      case "UUG" => "Leucine"
      case "UCU" => "Serine"
      case "UCC" => "Serine"
      case "UCA" => "Serine"
      case "UCG" => "Serine"
      case "UAU" => "Tyrosine"
      case "UAC" => "Tyrosine"
      case "UGU" => "Cysteine"
      case "UGC" => "Cysteine"
      case "UGG" => "Tryptophan"
      case "UAA" => "STOP"
      case "UAG" => "STOP"
      case "UGA" => "STOP"
      case _ => throw new IllegalArgumentException("Unknown codon")
    }
}

File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let proteins (rna: string): string list =
    let doProteins (rna: string): (string * string) option =
        match rna[0..2] with
        | "AUG" ->                         Some ("Methionine",    rna[3..])
        | "UGG" ->                         Some ("Tryptophan",    rna[3..])        
        | "UUC" | "UUU" ->                 Some ("Phenylalanine", rna[3..])
        | "UUA" | "UUG" ->                 Some ("Leucine",       rna[3..])
        | "UAU" | "UAC" ->                 Some ("Tyrosine",      rna[3..])
        | "UGU" | "UGC" ->                 Some ("Cysteine",      rna[3..])
        | "UCU" | "UCC" | "UCA" | "UCG" -> Some ("Serine",        rna[3..])
        | "UAA" | "UAG" | "UGA" | "" ->    None
        | _ -> failwith "Unknown coding"

    List.unfold doProteins rna